staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/vault.go

```diff
@@ -0,0 +1,218 @@
+	// Vault server SSL certificate.
+	CACert string `json:"ca-cert"`
+
+	// TLSServerName, if set, is used to set the SNI host when connecting via TLS.
+	TLSServerName string `json:"tls-server-name"`
+
+	// The path for transit API, default is "transit"
+	TransitPath string `json:"transit-path"`
+
+	// The path for auth backend, default is "auth"
+	AuthPath string `json:"auth-path"`
+}
+
+// KMSFactory function creates Vault KMS service
+func KMSFactory(configFile io.Reader) (envelope.Service, error) {
+	configFileContents, err := ioutil.ReadAll(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not read contents: %v", err)
+	}
+
+	var config EnvelopeConfig
+	err = yaml.Unmarshal(configFileContents, &config)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing file: %v", err)
+	}
+
+	err = checkConfig(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := newClientWrapper(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &vaultEnvelopeService{config: &config, client: client}, nil
+}
+
+func checkConfig(config *EnvelopeConfig) error {
+	if len(config.KeyNames) == 0 {
+		return errors.New("vault provider has no valid key names")
+	}
+
+	if config.Address == "" {
+		return errors.New("vault provider has no valid address")
+	}
+
+	return checkAuthConfig(config)
+}
+
+func checkAuthConfig(config *EnvelopeConfig) error {
+	var count uint
+
+	if config.Token != "" {
+		count++
+	}
+
+	if config.ClientCert != "" || config.ClientKey != "" {
+		if config.ClientCert == "" || config.ClientKey == "" {
+			return errors.New("vault provider has invalid TLS authentication information")
+		}
+		count++
+	}
+
+	if config.RoleID != "" || config.SecretID != "" {
+		if config.RoleID == "" {
+			return errors.New("vault provider has invalid approle authentication information")
+		}
+		count++
+	}
+
+	if count == 0 {
+		return errors.New("vault provider has no authentication information")
+	}
+	if count > 1 {
+		return errors.New("vault provider has more than one authentication information")
+	}
+
+	return nil
+}
+
+type vaultEnvelopeService struct {
+	config *EnvelopeConfig
+	client *clientWrapper
+	// We may update token for api.Client, but there is no sync for api.Client.
+	// Read lock for encrypt/decrypt requests, write lock for login requests which
+	// will update token for api.Client.
+	rwmutex sync.RWMutex
+}
+
+func (s *vaultEnvelopeService) Decrypt(data string) ([]byte, error) {
+	// Find the mached key
+	var key string
+	for _, name := range s.config.KeyNames {
+		if strings.HasPrefix(data, name+":") {
+			key = name
+			break
+		}
+	}
+	if key == "" {
+		return nil, errors.New("no matching vault key found")
+	}
+
+	// Replace the key name with "vault:" for Vault transit API
+	cipher := strings.Replace(data, key, vaultPrefix, 1)
+
+	//plain, _, err := s.client.decrypt(key, cipher)
+	plain, err := s.withRefreshToken((*clientWrapper).decrypt, key, cipher)
```

Comment:
It is a method defined in clientWrapper in client.go. it is just being called on clientWrapper object, but not directly from here, but from a more generic factored code in  method  withRefreshToken()

Reply:
Okay so I think you're doing this https://play.golang.org/p/LX0S1AWrM6

I don't think Kubernetes uses this pattern anywhere. I'd be a lot clearer if decrypt just took a clientWrapper as a first argument instead of a receiver.

Reply:
decrypt and encrypt interface are public interfaces defined already in the framework package: k8s.io/apiserver/pkg/storage/value/encrypt/envelope. This PR is just providing the implementation. If goal is to avoid the use of passing function , other option is to input a boolean in withRefreshToken() and in withRefreshToken() decide which method (encrypt or decrypt) to call based on boolean value.

Reply:
uploaded the changed code.